File: 1erdepartamental/python/7.1.identificar_tamanios.py
# ============================================
# PROGRAMA: Identificar Tamaños de Datos
# DESCRIPCIÓN: Analiza memoria de diferentes tipos
# ============================================

import sys


class AnalizadorTamanios:
    """
    Analiza tamaños en bytes de tipos Python.

    sys.getsizeof() incluye overhead del objeto
    """

    def __init__(self):
        self.__tipos = {}

    @property
    def tipos(self):
        """Retorna copia del diccionario"""
        return self.__tipos.copy()

    def analizar_basicos(self):
        """Mide tipos primitivos de Python"""
        self.__tipos["int"] = sys.getsizeof(0)
        self.__tipos["float"] = sys.getsizeof(0.0)
        self.__tipos["bool"] = sys.getsizeof(True)
        self.__tipos["str"] = sys.getsizeof("")
        self.__tipos["list"] = sys.getsizeof([])
        self.__tipos["dict"] = sys.getsizeof({})

    def mostrar(self):
        for tipo, tam in self.__tipos.items():
            print(f"{tipo}: {tam} bytes")


class Persona:
    """Clase simple para medir objetos personalizados"""

    def __init__(self, nombre, edad):
        self.__nombre = nombre
        self.__edad = edad

    @property
    def nombre(self):
        return self.__nombre

    @property
    def edad(self):
        return self.__edad


class TamanioObjeto:
    """Wrapper para sys.getsizeof()"""

    def __init__(self, obj):
        self.__obj = obj

    def obtener_tamanio(self):
        """Solo mide el contenedor, no objetos referenciados"""
        return sys.getsizeof(self.__obj)


# ============================================
# EJECUCIÓN
# ============================================
if __name__ == "__main__":
    a = AnalizadorTamanios()
    a.analizar_basicos()
    a.mostrar()

    print()
    p = Persona("Juan", 25)
    t = TamanioObjeto(p)
    print(f"Objeto Persona: {t.obtener_tamanio()} bytes")

File: 1erdepartamental/c++/7.1.identificar_tamanios.cpp
// ============================================
// PROGRAMA: Análisis de Tamaños de Memoria
// DESCRIPCIÓN: Demuestra sizeof para tipos básicos y custom
// ============================================

#include <iostream>
#include <string>
using namespace std;


/**
 * Tipo custom para análisis de tamaño
 */
class Persona {
private:
    string __nombre;
    string __apellido;
    int __edad;

public:
    Persona(string nombre, string apellido, int edad)
        : __nombre(nombre), __apellido(apellido), __edad(edad) {}
    
    friend ostream& operator<<(ostream& os, const Persona& p) {
        os << p.__nombre << " " << p.__apellido;
        return os;
    }
};


/**
 * Analizador de tamaños de memoria
 */
class AnalizadorTamanios {
public:
    void analizarTiposBasicos() const {
        cout << "=== TIPOS BÁSICOS ===" << endl;
        cout << "int: " << sizeof(int) << " bytes" << endl;
        cout << "double: " << sizeof(double) << " bytes" << endl;
        cout << "char: " << sizeof(char) << " bytes" << endl;
        cout << "bool: " << sizeof(bool) << " bytes" << endl;
        cout << "float: " << sizeof(float) << " bytes" << endl;
        cout << "long: " << sizeof(long) << " bytes" << endl;
        cout << endl;
    }
    
    void analizarString() const {
        cout << "=== STRINGS ===" << endl;
        string vacio = "";
        string corto = "Hola";
        string largo = "Este es un string mucho más largo";
        
        cout << "string vacío: " << sizeof(vacio) << " bytes" << endl;
        cout << "string corto: " << sizeof(corto) << " bytes" << endl;
        cout << "string largo: " << sizeof(largo) << " bytes" << endl;
        cout << "Nota: sizeof(string) mide el objeto, no el contenido" << endl;
        cout << endl;
    }
    
    void analizarObjeto() const {
        cout << "=== OBJETO CUSTOM ===" << endl;
        Persona p("Hugo", "Dominguez", 19);
        
        cout << "Persona: " << sizeof(p) << " bytes" << endl;
        cout << "Composición: 2 strings + 1 int" << endl;
    }
};


// ============================================
// EJECUCIÓN
// ============================================
int main() {
    AnalizadorTamanios analizador;
    
    analizador.analizarTiposBasicos();
    analizador.analizarString();
    analizador.analizarObjeto();
    
    return 0;
}
